tools/maphash/index.js

```
@@ -0,0 +1,134 @@
+	process.exit(1);
+}
+
+//let text = fs.readFileSync('../../_maps/map_files/YogStation/YogStation.dmm', 'utf8');
+//fs.writeFileSync('out.dmm', new DMM(text).toString());
+
+console.log("Running maphash...");
+
+let files = [];
+let file_outputs = new Map();
+let slashes = ['|', '/', '-', '\\'];
+let slash_index = 0;
+console.log();
+
+let converted_file_cache = {};
+
+async function do_dir(dirname) {
+	let dir = await fs.readdir(dirname, {withFileTypes: true});
+	let promises = [];
+	for(let ent of dir) {
+		if(ent.name == ".git" || ent.name == "node_modules") continue; // ignore folders with lots of irrelevant files.
+		let ent_path = path.join(dirname, ent.name);
+		if(ent.isDirectory()) {
+			promises.push(do_dir(ent_path));
+		} else if(ent.name.endsWith(".dmm")) {
+			promises.push(fs.readFile(ent_path, 'utf8').then((file_text) => {
+				process.stdout.write('\r' + slashes[(++slash_index) % 4]);
+				let hash;
+				if(converted_file_cache[path.resolve(ent_path)]) {
+					hash = crypto.createHash('sha1').update(file_text).digest('base64');
+					if(converted_file_cache[path.resolve(ent_path)] == hash) return;
+				}
+				try {
+					let dmm = new DMM(file_text);
+					let converted = dmm.toString();
+					if(converted != file_text) {
+						files.push(ent_path);
+						file_outputs.set(ent_path, converted);
+					} else {
+						if(!hash) hash = crypto.createHash('sha1').update(file_text).digest('base64');
+						converted_file_cache[path.resolve(ent_path)] = hash;
+					}
+				} catch(e) {
+					console.error("In " + ent_path);
+					console.error(e);
+				}
+			}));
+		}
+	}
+	await Promise.all(promises);
+}
+(async () => {
+	let initial_path = '../..';
+	if(process.argv.includes("--root-dir")) {
+		initial_path = process.argv[process.argv.indexOf('--root-dir') + 1];
+	}
+
+	try {
+		converted_file_cache = JSON.parse(await fs.readFile(path.join(__dirname, "converted_file_cache.json"), 'utf8'));
+	} catch(e) {}
```

Don't silently catch errors